emsdk.py

```diff
@@ -1245,8 +1245,34 @@ def is_optimizer_installed(tool):
   return os.path.exists(build_root)
 
 
-def build_optimizer_tool(tool):
-  debug_print('build_optimizer_tool(' + str(tool) + ')')
+# Finds newest tool of given name that is active, or if none are active, finds the newest one that is installed.
+def find_latest_active_or_installed_tool(name):
+  for t in reversed(tools):
+    if t.id == name and t.is_active():
+      return t
+
+  for t in reversed(tools):
+    if t.id == name and t.is_installed():
+      return t
+
+
+# npm install in Emscripten root directory
+def emscripten_npm_install(tool, directory):
+  node_tool = find_latest_active_or_installed_tool('node')
+  if not node_tool:
+    print('Failed to run "npm install" in installed Emscripten root directory ' + tool.installation_path() + '! Please install node.js first!')
+    return False
+
+  node_path = os.path.join(node_tool.installation_path(), 'bin')
+  npm = os.path.join(node_path, 'npm' + ('.cmd' if WINDOWS else ''))
+  env = os.environ.copy()
+  env["PATH"] = node_path + os.pathsep + env["PATH"]
```

Comment:
Just out of interest, why is this needed?

Reply:
Npm needs Node.js in PATH on Linux (on macOS and Windows it worked without).